### ISSUE TITLE
[Feature Request] Recursive root or multiple roots

### ISSUE THREAD

**Issue by sk-** [rubyatscale/stimpack#11]
At my current company, we have a bunch of existing code that is laid out in nested folders, which makes it difficult to start using `stimpack` as it only looks up packages in the first level of the root.

We see two alternatives that would work for us:
1. Change the behavior to search packs recursively
   We would need to change the [`resolve`](https://github.com/rubyatscale/stimpack/blob/12d1c0471823d4daee90b0f4cdcc7607d0af4b9a/lib/stimpack/packs.rb#L14-L21) method to something like:
   ```ruby
      def resolve
        # Gather all the packs under the root directory and create packs.
        Dir.glob(Rails.root.join(Stimpack.config.root).join("**").join("package.yml")).sort!.each do |path|
          next unless pack = Pack.create(File.dirname(path))
          @packs[pack.name] = pack
        end
        @packs.freeze
      end   
   ```

2. Allow having multiple config roots, in that way we could manually list all the folder containing our code.

Would you be willing to accept a change to support this?

**Comment by AlexEvanczuk** [comment]
@sk- Thanks for opening this issue. We'd love to take a look at a PR if you have an idea of how you'd implement this. One thing we discussed as a team was allowing the client to inject their own behavior for how packs are found and instantiated. My one concern is that a lot of the other parts of the ruby at scale ecosystem (both stimpack and other libraries) may/do depend no the flat structure to work properly, so my hesitancy is that we'll create subtle issues in other libraries if we don't apply this change carefully.

Let me know what ya think. I think we'd be curious to see the implementation you had in mind and what it might break in the larger ecosystem, and it would also be interesting to pursue an approach where you can inject your own resolution strategy.

**Comment by sk-** [comment]
Hi @AlexEvanczuk, thanks for the reply. Let me first give you some context about how the code is currently laid out in our organization.

Currently we have a large number of engines which are all present in the folder `engines` and are organized in different subfolders (18 in total), corresponding either to a product area or team.

Something like the following:

```
engines
|_ area1
    |__ engine_a -> namespace is EngineA
    |__ engine_b -> namespace is EngineB
|_ area2
    |__ engine_c -> namespace is EngineC
    |__ engine_d -> namespace is EngineD
```

some of those engines are ill conceived, in the sense that they are either too small or have circular dependencies, but it is really complex for us to changes due to the amount of code (~500k lines).

We want to start using stimpack, so we thought to convert them to engine packs. But at the same time we wanted to also allow to have the new packs in a new folder called `packs`, where we could have more strict settings.

So for our purposes we would need to have multiple roots, and also recursive discovery.

To implement the latter we tested the following https://github.com/sk-/stimpack/pulls. Not sure what the implications of doing a recursive pack resolution are.

Maybe a better alternative would be to still support a flat structure, but with multiple roots. In that way we could set the roots as:

```
engines/area1
engines/area2
packs                  # This would be the new shiny packs with a more strict config
```

Let me know what you think!

**Comment by AlexEvanczuk** [comment]
@sk- Thanks for sharing more info. We actually have our application structured in a similar way too. For us it's:
```
packs
components
  engine1
  engine2
  engine3
```

This works okay with us with the way stimpack works today. Can you share a bit more about the behavior that you are expecting that you aren't getting today? That would help me better understand the problem!

**Comment by sk-** [comment]
In our case we have the engines one level below (it would be `components -> product -> engine1`), so they won't be found by stimpack if we specify `components` as the `root`.

Also I'm curious how are you able to tell stimpack that you have both a `packs` and a `components` folder?

**Comment by AlexEvanczuk** [comment]
Ah gotcha. So -- we actually don't have stimpack read the components folder. Note that stimpack isn't needed for engines to be able to be packs. It adds features (like rspec integration, autoload path setup, etc.) that we don't actually need for our engines. What features of stimpack are you hoping to have for your engines?

**Comment by sk-** [comment]
In our case, we want to get rid of the engines boilerplate and use the `engine` metadata (https://github.com/rubyatscale/stimpack/discussions/14#discussioncomment-3242722).

**Comment by AlexEvanczuk** [comment]
Thanks for sharing — it makes sense to want to shed that boiler plate. I chatted with my team and I don’t think we’ll be able to add support for multiple roots right now as we won’t be able to commit to maintaining it (both in stimpack and its downstream implications across the rubyatscale ecosystem). It seems to me like using stimpack to shed the boiler plate is not blocking stimpack adoption and its a nice-to-have on top of stimpack adoption. If this is the case, then another approach could be to adopt stimpack with the ./packs directory as normal and keep the engines as they are, or alternatively to move engines into that same ./packs directory, which is where we moved our engines that we to shed the boiler plate for. Let me know what you think or if you have any other concerns.

**Comment by sk-** [comment]
But then, could you consider having just one root but searching the packs recursively. Otherwise one would still have the issue of having too many packs in the same folder. We currently have more than 90 engines, and if we were to put all of them in the same `packs` folder it would be quite hard to maintain.

The change to support a recursive root is pretty straightforward.

**Comment by AlexEvanczuk** [comment]
We plan to add support for (one-level deep) nested packages soon, which will look like this:
```
packs/foo/package.yml
packs/foo/
  bar/package.yml
  baz/package.yml
```
You could arrange the engines like:
```
packs/domain1/package.yml
packs/domain2/package.yml
packs/engines/
  engine1/package.yml
  engine2/package.yml
  engine3/package.yml
```

Do you think this would resolve things for you?

We know that in stimpack alone supporting some of these different changes may be simpler mechanically, but one of our goals is we want to make sure the whole rubyatscale ecosystem supports major assumptions of other components so each component can be used with each other component regardless of how the client uses one of the supported configurations. (That being said, I think there is room to improve the architecture of the other systems so they are not implicitly coupled to specific assumptions of other systems, such as `package_protections` asking `stimpack` about roots.)

**Comment by sk-** [comment]
Thanks @AlexEvanczuk for bearing with me and explaining the rationale. That solution would work great for us.

We would be happy to contribute in any way in order to be able to ship this soon.

**Comment by AlexEvanczuk** [comment]
@sk- We JUST released 0.0.6 which adds support for nested packs in `stimpack`: https://github.com/rubyatscale/stimpack/pull/23

Note that the rest of the `rubyatscale` ecosystem doesn't yet support it (we are planning to add support for nested packs in the rest of the ecosystem next week). In the meantime, since I imagine you're just getting your feet wet with stimpack and aren't using other tools in the ecosystem, feel free to give this a try and let me know if you have any feedback!!

**Comment by oleg-vinted** [comment]
Related issue: https://github.com/rubyatscale/stimpack/issues/26.
